grpclog/component.go

```diff
@@ -0,0 +1,289 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package grpclog
+
+import (
+	"fmt"
+	"math"
+	"os"
+	"strconv"
+	"strings"
+
+	"google.golang.org/grpc/internal/grpclog"
+)
+
+// Component is a unique identifier for a component. Created with NewComponent.
+type Component string
+
+// componentData records the settings for a component.
+type componentData struct {
+	verbosity int
+	level     int
+}
+
+const envName = "GRPC_GO_LOG_LEVEL"
+const sentinel = math.MinInt32
+
+// grpc components
+const (
+	CORE       = Component("CORE")
+	ALTS       = Component("ALTS")
+	GRPCLB     = Component("BALANCER_GRPCLB")
+	RLS        = Component("BALANCER_RLS")
+	ROUNDROBIN = Component("BALANCER_ROUNDROBIN")
+	BINARYLOG  = Component("BINARYLOG")
+	CHANNELZ   = Component("CHANNELZ")
+	DNS        = Component("RESOLVER_DNS")
+	TRANSPORT  = Component("TRANSPORT")
+	XDS        = Component("XDS")
+)
+
+var components = map[Component]*componentData{}
+var grpcComponents = []Component{
+	CORE,
+	ALTS,
+	GRPCLB,
+	RLS,
+	ROUNDROBIN,
+	BINARYLOG,
+	CHANNELZ,
+	DNS,
+	TRANSPORT,
+	XDS,
+}
+var environmentVars = map[string]*componentData{}
+var prefixVars = map[string]*componentData{}
+
+// init creates all grpc components and any components named in the environment variable and applies all settings
+// specified in the environment variable. Notably, settings for all components with a given prefix are applies before
+// regular component settings.
+func init() {
+	// Initialize the grpc components
+	for _, c := range grpcComponents {
+		components[c] = &componentData{0, 0}
+	}
+	// Pull environment variable data
+	if envVar, ok := os.LookupEnv(envName); ok && len(envVar) > 0 {
+		varList := strings.Split(envVar, ",")
+		for _, varPair := range varList {
+			varPairList := strings.Split(varPair, ":")
+			if len(varPairList) != 2 {
+				fmt.Fprintf(os.Stderr, "error: could not parse %v value '%v', unrecognized key-value pair '%v'\n", envName, envVar, varPair)
+				os.Exit(1)
+			}
+			if cData, ok := parseVar(varPairList[1]); ok {
+				if varPrefix, ok := getPrefix(varPairList[0]); ok {
+					prefixVars[varPrefix] = &componentData{cData.verbosity, cData.level}
+				} else {
+					environmentVars[varPairList[0]] = &componentData{cData.verbosity, cData.level}
+				}
+			} else {
+				fmt.Fprintf(os.Stderr, "error: could not parse %v value '%v', unrecognized value '%v'\n", envName, envVar, varPairList[1])
+				os.Exit(1)
+			}
+		}
+	}
+	// Create any components named in the environment variable that do not exist
+	for name := range environmentVars {
+		c := Component(name)
+		if _, ok := components[c]; !ok {
+			components[c] = &componentData{0, 0}
+		}
+	}
+	// Apply prefixes
+	for prefix, pData := range prefixVars {
+		for c, cData := range components {
+			if strings.HasPrefix(string(c), prefix) {
+				cData.apply(pData)
```

Comment:
I think component (exact match) should take precedence over prefix (partial) match.

Reply:
Agreed. Prefix settings are applied before the more specific component settings so that the latter overrides the former (unless I am mistaken).

Reply:
Oh, sorry, I missed the apply after this.